public/app/core/utils/explore.ts

```diff
@@ -254,8 +241,34 @@ export function generateKey(index = 0): string {
   return `Q-${uuidv4()}-${index}`;
 }
 
-export function generateEmptyQuery(queries: DataQuery[], index = 0): DataQuery {
-  return { refId: getNextRefIdChar(queries), key: generateKey(index) };
+export async function generateEmptyQuery(
+  queries: DataQuery[],
+  index = 0,
+  dataSourceOverride?: DataSourceRef
+): Promise<DataQuery> {
+  let datasourceFull;
+  let datasourceRef;
+  let defaultQuery;
+
+  // datasource override is if we have switched datasources with no carry-over - we want to create a new query with a datasource we define
+  if (dataSourceOverride) {
+    datasourceRef = dataSourceOverride;
+  } else if (queries.length > 0 && queries[queries.length - 1].datasource) {
+    // otherwise use last queries' datasource
+    datasourceRef = queries[queries.length - 1].datasource;
+  } else {
+    // if neither exists, use the default datasource
+    datasourceFull = await getDataSourceSrv().get();
+    defaultQuery = datasourceFull.getDefaultQuery?.(CoreApp.Explore);
+    datasourceRef = datasourceFull.getRef();
+  }
+
+  if (!datasourceFull) {
+    datasourceFull = await getDataSourceSrv().get(datasourceRef);
+    defaultQuery = datasourceFull.getDefaultQuery?.(CoreApp.Explore);
+  }
```

Comment:
just to not have implicit `any`s around:

```suggestion
  let datasourceInstance: DataSourceApi | undefined;
  let datasourceRef: DataSourceRef | null | undefined;
  let defaultQuery: Partial<DataQuery> | undefined;

  // datasource override is if we have switched datasources with no carry-over - we want to create a new query with a datasource we define
  if (dataSourceOverride) {
    datasourceRef = dataSourceOverride;
  } else if (queries.length > 0 && queries[queries.length - 1].datasource) {
    // otherwise use last queries' datasource
    datasourceRef = queries[queries.length - 1].datasource;
  } else {
    // if neither exists, use the default datasource
    datasourceInstance = await getDataSourceSrv().get();
    defaultQuery = datasourceInstance.getDefaultQuery?.(CoreApp.Explore);
    datasourceRef = datasourceInstance.getRef();
  }

  if (!datasourceInstance) {
    datasourceInstance = await getDataSourceSrv().get(datasourceRef);
    defaultQuery = datasourceInstance.getDefaultQuery?.(CoreApp.Explore);
  }
```